### PR TITLE
feat(dhall-docs): add jump-to-definition support for record-like expressions

### PR DESCRIPTION
**Pull request by german1608**
(this is a quite big PR, please read the full description of this PR before reviewing the code)

### Summary

This PR adds JTD support for record-like expressions. `dhall-docs` now tries to infer (really simple, not like `Dhall.TypeCheck` does, for example) the type of a dhall expression. As a result of this, the JTD-related features on record-like expressions are:

* Fields on field-access expressions jumps to the definition of that field in the source code, where the left side of that expression was defined;
* Deep field expressions are also supported (see `JumpToRecordFieldWhenVarIsOfRecordTypeDeep` test case);
* Fields that use dot-syntax are correctly highlighted depending on the accessed field;
* Punned entries are handled without jumping to the definition of that entry, but it's enough I think;
* It supports transitivity of expressions, which allows jumping to the _real_ definition of an expression if `dhall-docs` is able to keep the inferred type in a transitivity chain.

You can check the written test cases to see more specific uses, and also you can suggest more test cases as well.

### How to review this

This work is divided into two parts:

1. Modifications of how we preserve whitespace on the AST:
    1. The addition of the `FieldAccess` data-type for preserving whitespace on `Field` expression
    2. The correct whitespace preservation on `dot-syntax` labels, which is documented in the `RecordField` haddock
    3. Modify the rest of the repository packages to catch-up the change on `Field`
2. Modifications on how the source code on rendered in `fragments`:
    * We have a `DhallType` ADT to aim a simple type-inference algorithm.
    * Now we use a `Writer [SourceCodeFragment] DhallType`. Generating `[SourceCodeFragment]` for an expression will return the inferred 'DhallType'

The part 1 generates a lot of noise on the PR, but I was modifying the parser as I was developing record-JTD since some changes were really specific for `dhall-docs`. If you need to, I can split (not so easy, though) the `dhall` work in a separate PR that should be merged before this one.

### CODE REVIEW

File: dhall-docs/tasty/data/package/JumpToRecordFieldWhenVarIsOfRecordTypeTransitivity.dhall [hunk 1]
```diff
@@ -0,0 +1,17 @@
+{-|
+On this example, `dhall-docs` will detect that the variable `a` is bounded
+to a record literal and it will be aware that it uses dot-syntax.
+
+Transitivity is respected when the variable is assigned using let-bindings.
+For instance, if the user clicks the `y` from the `ax.y.z`, it will jump to
+the definition of `x` on the record literal.
+
+_(the code is the same from <./JumpToRecordFieldWhenVarIsOfRecordTypeWithDotSyntax.dhall>)_
```
Comment (Gabriel439) [inline dhall-docs/tasty/data/package/JumpToRecordFieldWhenVarIsOfRecordTypeTransitivity.dhall:9]: What is the reason for having duplicate examples?

File: dhall-docs/tasty/data/package/JumpToRecordFieldWhenVarIsAnnotatedWithRecordType.dhall [hunk 2]
```diff
@@ -0,0 +1,11 @@
+{-|
+On this example, `dhall-docs` will detect that the variable `a` is annotated
+with a record-type, and associate all jump-to-definition features on its field
+access to jump to the fields on that record-type
+
+Note that the `x` field on the record literal isn't highlighted since we consider
+that the variable was _defined_ in the annotation
```
Comment (Gabriel439) [inline dhall-docs/tasty/data/package/JumpToRecordFieldWhenVarIsAnnotatedWithRecordType.dhall:7]: It seems like we should prefer the record literal over the record type, when both are present

File: dhall-docs/tasty/data/package/JumpToRecordFieldWhenVarIsOfRecordTypeTransitivity.dhall (hunk 1 above)
Comment (german1608) [reply dhall-docs/tasty/data/package/JumpToRecordFieldWhenVarIsOfRecordTypeTransitivity.dhall:9]: I like to keep one feature or bug-fix per file. Although the file is a duplicate, it is a good example of transitivity.

We can change the sample, if you want to

File: dhall-docs/tasty/data/package/JumpToRecordFieldWhenVarIsAnnotatedWithRecordType.dhall (hunk 2 above)
Comment (german1608) [reply dhall-docs/tasty/data/package/JumpToRecordFieldWhenVarIsAnnotatedWithRecordType.dhall:7]: Then the annotation should not affect anything. Is this what you suggest?

Comment (Gabriel439) [reply dhall-docs/tasty/data/package/JumpToRecordFieldWhenVarIsAnnotatedWithRecordType.dhall:7]: Yes, that's correct

File: dhall-docs/tasty/data/package/JumpToRecordFieldWhenVarIsOfRecordTypeTransitivity.dhall (hunk 1 above)
Comment (Gabriel439) [reply dhall-docs/tasty/data/package/JumpToRecordFieldWhenVarIsOfRecordTypeTransitivity.dhall:9]: I think at least minimize each example to the smallest code that illustrates what's going on

File: dhall/src/Dhall/Parser/Expression.hs [hunk 3]
```diff
@@ -443,18 +444,29 @@ parsers embedded = Parsers {..}
...
+            let fieldParser = do
+                    src0 <- src whitespace
+                    l <- anyLabel
+                    pos <- getSourcePos
+
+                    -- FIXME: Suffix whitespace can't be parsed given our limitation
+                    -- about whitespace treatment, but for @dhall-docs@ this
+                    -- is enough
+                    let src1 = Src pos pos ""
+                    return (FieldAccess (Just src0) l (Just src1))
```
Comment (sjakobi) [inline dhall/src/Dhall/Parser/Expression.hs:462]: I don't understand this FIXME.  `anyLabel` doesn't eat trailing whitespace, or does it?!

I think `src1` should properly include all the whitespace until the next `.`.

I also think we direly need tests for the whitespace parsing – possibly golden tests based on `haskell-syntax-tree`?! That's for future work though, I guess.

File: dhall/src/Dhall/Normalize.hs [hunk 4]
```diff
@@ -891,7 +896,7 @@ isNormalized e0 = loop (Syntax.denote e0)
       ToMap x t -> case x of
           RecordLit _ -> False
           _ -> loop x && all loop t
-      Field r k -> case r of
+      Field r (Syntax.fieldAccessLabel -> k) -> case r of
```
Comment (sjakobi) [inline dhall/src/Dhall/Normalize.hs:899]: Since normalization turns all the `Maybe s` fields into `Nothing`, I think we need to check that these fields are `Nothing`. We should do the same for `RecordField` too.

File: dhall/tests/Dhall/Test/Dhall.hs [hunk 5]
```diff
@@ -116,9 +116,10 @@ shouldTreatAConstructorStoringUnitAsEmptyAlternative = testCase "Handle unit con
     let exampleEncoder :: Dhall.Encoder ()
         exampleEncoder = Dhall.unionEncoder (Dhall.encodeConstructor "A")
 
-    Dhall.embed exampleEncoder () @=? Field (Union (Dhall.Map.singleton "A" Nothing)) "A"
+    let a = Dhall.Core.makeFieldAccess "A"
+    Dhall.embed exampleEncoder () @=? Field (Union (Dhall.Map.singleton "A" Nothing)) a
 
-data RecursiveType a = RecursiveType (RecursiveType a)
+newtype RecursiveType a = RecursiveType (RecursiveType a)
```
[the remark this answers is not in the document]
Comment (german1608) [reply dhall/tests/Dhall/Test/Dhall.hs:122]: I changed that manually, I have a TOC that males me change to newtype when possible :) I can rollback the change, though

**Comment by german1608** [comment]
> @german1608, can you show us some docs that show off the new features?

I'm unable to provide specific examples right now since I'm not home, but you can check the generated docs for the test package and the prelude. renderAs on Prelude/JSON has a lot of stuff

File: dhall-docs/tasty/data/package/JumpToRecordFieldWhenVarIsAnnotatedWithRecordType.dhall (hunk 2 above)
Comment (german1608) [reply dhall-docs/tasty/data/package/JumpToRecordFieldWhenVarIsAnnotatedWithRecordType.dhall:7]: OK, on let-bindings we ignore the type from the annotation on this type-inference

OTOH I believe that in the case of function arguments of record-type from `Lam`, selector expressions should highlight the fields on the record-type. This could be useful to get the associated type of that field in the selector expression when the record-type is huge.

Check https://hydra.dhall-lang.org/build/69827/download/1/docs/JSON/object.dhall.html for example,  hovering the `object` in `json.object` results in the highlight of the `object` field in the record-type, which is really useful IMO

File: dhall/src/Dhall/Parser/Expression.hs (hunk 3 above)
Comment (german1608) [reply dhall/src/Dhall/Parser/Expression.hs:462]: I had a lot of parsing errors from our test-suite when I added the `src1 <- src whitespace` there, I can try to reproduce and paste a sample of the failure I got to get more help from you and @Gabriel439

I agree with the test-setup you suggest, I'll file an issue to not forget about it

**Comment by german1608** [comment]
> > @german1608, can you show us some docs that show off the new features?
> 
> I'm unable to provide specific examples right now since I'm not home, but you can check the generated docs for the test package and the prelude. renderAs on Prelude/JSON has a lot of stuff

On the `Prelude` there isn't a lot of examples, some JTD features can be seen in the [`Prelude/JSON` subpackage](https://hydra.dhall-lang.org/build/69827/download/1/docs/JSON/index.html).

The following examples on the test packages shows off the new features:
* https://hydra.dhall-lang.org/build/69815/download/1/docs/JumpToRecordFieldWhenFieldDoesNotExist.dhall.html
* https://hydra.dhall-lang.org/build/69815/download/1/docs/JumpToRecordFieldWhenVarIsAnnotatedWithNonRecordType.dhall.html
* https://hydra.dhall-lang.org/build/69815/download/1/docs/JumpToRecordFieldWhenVarIsAnnotatedWithRecordType.dhall.html
* https://hydra.dhall-lang.org/build/69815/download/1/docs/JumpToRecordFieldWhenVarIsOfRecordType.dhall.html
* https://hydra.dhall-lang.org/build/69815/download/1/docs/JumpToRecordFieldWhenVarIsOfRecordTypeComplex.dhall.html
* https://hydra.dhall-lang.org/build/69815/download/1/docs/JumpToRecordFieldWhenVarIsOfRecordTypeDeep.dhall.html
* https://hydra.dhall-lang.org/build/69815/download/1/docs/JumpToRecordFieldWhenVarIsOfRecordTypeOnPunnedEntry.dhall.html
* https://hydra.dhall-lang.org/build/69815/download/1/docs/JumpToRecordFieldWhenVarIsOfRecordTypeTransitivity.dhall.html
* https://hydra.dhall-lang.org/build/69815/download/1/docs/JumpToRecordFieldWhenVarIsOfRecordTypeWithDotSyntax.dhall.html

...or you can manually check on the [index package](https://hydra.dhall-lang.org/build/69815/download/1/docs/index.html) on all the examples that starts with `JumpToRecordFieldWhen...`

File: dhall-docs/tasty/data/package/JumpToRecordFieldWhenVarIsAnnotatedWithRecordType.dhall (hunk 2 above)
Comment (Gabriel439) [reply dhall-docs/tasty/data/package/JumpToRecordFieldWhenVarIsAnnotatedWithRecordType.dhall:7]: @german1608: Yes, I agree that function argument types should be highlighted when possible.  I was only commenting on `let` expressions

Comment (german1608) [reply dhall-docs/tasty/data/package/JumpToRecordFieldWhenVarIsAnnotatedWithRecordType.dhall:7]: I'm adding a specific test for `Lam` bindings to reflect that behavior and updating the existing test case for `Let` bindings as well

File: dhall/src/Dhall/Parser/Expression.hs (hunk 3 above)
Comment (german1608) [reply dhall/src/Dhall/Parser/Expression.hs]: Alright, after replacing the `let src1 = Src pos pos ""` with `src1 <- src whitespace` I got a lot of parse errors from the test-suite. This is a single test failure:

```
Dhall Tests
  normalization
    discover
      ./dhall-lang/tests/normalization/success/remoteSystems:                                                                  FAIL
        Exception: 
        [1;31mError[0m: Invalid input
        
        41:3:
           |
        41 |   (   λ ( x
           |   ^
        unexpected '('
        expecting ->, :, end of input, or whitespace
```

the error comes from parsing https://github.com/dhall-lang/dhall-lang/blob/master/tests/normalization/success/remoteSystemsB.dhall

do you have any idea? @Gabriel439 may have a solution or more insights about the reason for the failure.

File: dhall/src/Dhall/Parser/Expression.hs [hunk 6]
```diff
@@ -443,18 +444,29 @@ parsers embedded = Parsers {..}
...
             a <- primitiveExpression
 
-            let recordType = _openParens *> whitespace *> expression <* whitespace <* _closeParens
+            let recordType = whitespace *> _openParens *> whitespace *> expression <* whitespace <* _closeParens
 
             let field               x  e = Field   e  x
             let projectBySet        xs e = Project e (Left  xs)
             let projectByExpression xs e = Project e (Right xs)
 
+            let fieldParser = do
```
Comment (sjakobi) [inline dhall/src/Dhall/Parser/Expression.hs:453]:
```suggestion
            let fieldSelection = do
```

File: dhall/src/Dhall/Syntax.hs [hunk 7]
```diff
@@ -289,33 +291,70 @@ instance Bifunctor PreferAnnotation where
...
+will be instantiated as follows:
+
+* @recordFieldSrc0@ corresponds to the @A@ comment.
+* @recordFieldValue@ corresponds to @(Var "x")@
+* @recordFieldSrc1@ corresponds to the @B@ comment.
+* @recordFieldSrc2@ will be 'Nothing'
+
+If dot-syntax is being used all three srcs are 'Just's, but for
+every label except the last one @recordFieldSrc1@ should be equal to
+@recordFieldSrc2@. On the last label, @recordFieldSrc2@ semantics are the same
```
Comment (sjakobi) [inline dhall/src/Dhall/Syntax.hs:324]: Why do we have this duplication? Why can't `recordFieldSrc2` simply be `Nothing` when the label isn't directly followed by a `=`?

File: dhall/src/Dhall/Syntax.hs [hunk 8]
```diff
@@ -289,33 +291,70 @@ instance Bifunctor PreferAnnotation where
...
+as described from above. Also, for any subsequent labels l1 and l2,
+@recordFieldSrc1 l1 == recordFieldSrc0 l2@. For example,
```
Comment (sjakobi) [inline dhall/src/Dhall/Syntax.hs:326]: This doesn't seem to be the case for e.g. `a` and `b` in the example below.

File: dhall/src/Dhall/Normalize.hs [hunk 9]
```diff
@@ -891,7 +905,7 @@ isNormalized e0 = loop (Syntax.denote e0)
       ToMap x t -> case x of
           RecordLit _ -> False
           _ -> loop x && all loop t
-      Field r k -> case r of
+      Field r (FieldSelection s0 k s1) -> Maybe.isNothing s0 && Maybe.isNothing s1 && case r of
```
Comment (sjakobi) [inline dhall/src/Dhall/Normalize.hs:908]: I think it would be slightly more readable to handle the `Maybe s` fields with two pattern matches, where the first is
```suggestion
      Field r (FieldSelection Nothing k Nothing) -> case r of
```

File: dhall/src/Dhall/Parser/Expression.hs (hunk 3 above)
Comment (sjakobi) [reply dhall/src/Dhall/Parser/Expression.hs]: Thanks for investigating, @german1608! I'm fine with leaving this bit as it is for now then.

**Comment by german1608** [comment]
> In https://hydra.dhall-lang.org/build/69960/download/1/docs/JSON/renderAs.dhall.html I noticed that every single label in a record literal seems to be highlighted. But when I hover over these or click them, nothing interesting happens. What's the reason for highlighting these labels?

That is because the implementation doesn't recognize where a record (literal or type) is stored and in any case it renders the highlight. On this moment, we recognize when a record literal has been bound to a let binding and when a lambda name is of record-type, but the rest of the cases are not handled.

I guess that we could remove the highlight on unused names, although when handling imported names (as I explained [here](https://github.com/dhall-lang/dhall-haskell/issues/1994#issuecomment-674080303)) it would get messy. Something that a couple of tests doesn't solve.

File: dhall/src/Dhall/Syntax.hs (hunk 7 above)
Comment (german1608) [reply dhall/src/Dhall/Syntax.hs:324]: You're right. My rationale was to have a way to differ dot-syntax from punned entries, but I didn't notice that by doing this we don't have a way to differ between dot-syntax and normal labels

File: dhall/src/Dhall/Parser/Expression.hs (hunk 3 above)
Comment (german1608) [reply dhall/src/Dhall/Parser/Expression.hs]: You're welcome! I think that I'll update #1995 to not forget about this

**Comment by german1608** [comment]
I'll be quite bussy today, so I'll address remaining comments this night.

File: dhall/src/Dhall/Syntax.hs (hunk 7 above)
Comment (german1608) [reply dhall/src/Dhall/Syntax.hs:324]: I changed the semantics as you suggested, @sjakobi!

**Comment by sjakobi** [comment]
> I guess that we could remove the highlight on unused names, although when handling imported names (as I explained [here](https://github.com/dhall-lang/dhall-haskell/issues/1994#issuecomment-674080303)) it would get messy. Something that a couple of tests doesn't solve.

OK. Currently it seems to me that there's not much point in highlighting things that don't offer more information in the same view. So when a label in one file is only used in other files, I think it's probably not worth highlighting in the first file.

Maybe this issue could also be addressed by using a more sophisticated highlighting approach that uses different visual cues instead of the current one-size-fits-all design with the underlining. I'm not sure.

I noticed that some types in the index pages also contain highlighting, e.g. in https://hydra.dhall-lang.org/build/70067/download/1/docs/List/index.html. Can these labels be jumped to from anywhere?

File: dhall-docs/src/Dhall/Docs/CodeRenderer.hs [hunk 10]
```diff
@@ -103,68 +125,119 @@ data SourceCodeFragment =
...
+            let match (VarDecl _ l _) = l == label
+            case filter match fields of
+                x@(VarDecl _ _ t) : _ -> do
+                    Writer.tell [SourceCodeFragment src (VariableUse x)]
+                    return t
+                _ -> return AnyType
 
+        RecordLit (Map.toList -> l) -> handleRecordLike l
+
+        Record (Map.toList -> l) -> handleRecordLike l
```
Comment (sjakobi) [inline dhall-docs/src/Dhall/Docs/CodeRenderer.hs:205]: In which cases is it useful that we record a `RecordLiteral` for a `Record` _type_?

(I also wonder whether `RecordLiteral` should have a better name.)

File: dhall-docs/src/Dhall/Docs/CodeRenderer.hs [hunk 11]
```diff
@@ -68,12 +77,25 @@ getSourceLine, getSourceColumn :: SourcePos -> Int
 getSourceLine = SourcePos.unPos . SourcePos.sourceLine
 getSourceColumn = SourcePos.unPos . SourcePos.sourceColumn
 
+-- | To support jump-to-definition on record literals we need to know the
+--   the type of a name as we traverse the AST. Using 'Dhall.TypeCheck' is not
+--   possible since we always try to extract as much information as we can from
+--   source.
+data DhallType
```
Comment (sjakobi) [inline dhall-docs/src/Dhall/Docs/CodeRenderer.hs:84]: I was initially confused in what way this represents a "type". I wonder whether something like `JtdInfo` or maybe `Provenience` might be slightly less confusing.

**Comment by german1608** [comment]
> OK. Currently it seems to me that there's not much point in highlighting things that don't offer more information in the same view. So when a label in one file is only used in other files, I think it's probably not worth highlighting in the first file.


I see. Then let me do some work to remove the unused highlights. It shouldn't be difficult

File: dhall-docs/src/Dhall/Docs/CodeRenderer.hs (hunk 11 above)
Comment (german1608) [reply dhall-docs/src/Dhall/Docs/CodeRenderer.hs:84]: I'll use `JtdInfo`. `Provenience` doesn't seem the best term from my POV.

File: dhall-docs/src/Dhall/Docs/CodeRenderer.hs (hunk 10 above)
Comment (german1608) [reply dhall-docs/src/Dhall/Docs/CodeRenderer.hs:205]: > (I also wonder whether `RecordLiteral` should have a better name.)

This is the problem. In the beginning, I used it only for record literals, and later I added support for `Record` using the same name.

I'll use `RecordFields`: generic and more appropriate with the `JtdInfo` type name you suggested

**Comment by german1608** [comment]
@sjakobi Alright, I've applied your suggestions and I removed the highlighting on unused names. Note that unused let-bindings are not highlighted as well. Check [this](https://hydra.dhall-lang.org/build/70134/download/1/docs/JumpToDefOnUnused.dhall.html) test case

A thing that I'd like to do is to rename `var` to `name` when possible since I feel that `var` is more specific to `let-bindings`, but I'll do it in a follow-up PR to not add more noise here

File: dhall-docs/src/Dhall/Docs/CodeRenderer.hs [hunk 12]
```diff
@@ -68,12 +78,42 @@ getSourceLine, getSourceColumn :: SourcePos -> Int
...
+
+        > { a = foo, b = bar }
+
+        has the following 'JtdInfo':
+
+        > RecordFields (Set.fromList [VarDecl posA "a" jtdInfoA, VarDecl posB "b" jtdInfoB])
+
+        ... where
+
+        * @posA@ and @posB@ records the source position used to make them
```
Comment (sjakobi) [inline dhall-docs/src/Dhall/Docs/CodeRenderer.hs:100]:
```suggestion
        * @posA@ and @posB@ record the source position used to make them
```

File: dhall-docs/src/Dhall/Docs/CodeRenderer.hs [hunk 13]
```diff
@@ -68,12 +78,42 @@ getSourceLine, getSourceColumn :: SourcePos -> Int
...
+        identifiel across the rendered source code
```
Comment (sjakobi) [inline dhall-docs/src/Dhall/Docs/CodeRenderer.hs:101]:
```suggestion
        identified across the rendered source code
```

File: dhall-docs/src/Dhall/Docs/CodeRenderer.hs [hunk 14]
```diff
@@ -68,12 +78,42 @@ getSourceLine, getSourceColumn :: SourcePos -> Int
...
+        * @jtdInfoA@ and @jtdInfoB@ are the associated 'JtdInfo' infered from
```
Comment (sjakobi) [inline dhall-docs/src/Dhall/Docs/CodeRenderer.hs:102]:
```suggestion
        * @jtdInfoA@ and @jtdInfoB@ are the associated 'JtdInfo's inferred from
```

File: dhall-docs/src/Dhall/Docs/CodeRenderer.hs [hunk 15]
```diff
@@ -103,68 +143,133 @@ data SourceCodeFragment =
 -- | Returns all 'SourceCodeFragment's in lexicographic order i.e. in the same
 --   order as in the source code.
 fragments :: Expr Src Import -> [SourceCodeFragment]
-fragments = go Context.empty
+fragments = removeUnusedDecls . Data.List.sortBy sorter . Writer.execWriter . infer Context.empty
```
Comment (sjakobi) [inline dhall-docs/src/Dhall/Docs/CodeRenderer.hs:146]: Filtering before sorting should be slightly faster

```suggestion
fragments = Data.List.sortBy sorter . removeUnusedDecls . Writer.execWriter . infer Context.empty
```

**Comment by german1608** [comment]
> Great job, @german1608!

Thanks for your good reviews!
